### ISSUE TITLE
x/crypto/cryptobyte/asn1: support reading implicitly-tagged values

### ISSUE THREAD

**Issue by aarongable** [golang/go#64811]
### Go version

go version go1.21.5 linux/amd64

### What operating system and processor architecture are you using (`go env`)?

```shell
GO111MODULE=''
GOARCH='amd64'
GOBIN=''
GOCACHE='/home/aaron/.cache/go-build'
GOENV='/home/aaron/.config/go/env'
GOEXE=''
GOEXPERIMENT=''
GOFLAGS=''
GOHOSTARCH='amd64'
GOHOSTOS='linux'
GOINSECURE=''
GOMODCACHE='/home/aaron/.local/share/go/pkg/mod'
GONOPROXY=''
GONOSUMDB=''
GOOS='linux'
GOPATH='/home/aaron/.local/share/go'
GOPRIVATE=''
GOPROXY='https://proxy.golang.org,direct'
GOROOT='/usr/local/go'
GOSUMDB='sum.golang.org'
GOTMPDIR=''
GOTOOLCHAIN='auto'
GOTOOLDIR='/usr/local/go/pkg/tool/linux_amd64'
GOVCS=''
GOVERSION='go1.21.5'
GCCGO='gccgo'
GOAMD64='v1'
AR='ar'
CC='gcc'
CXX='g++'
CGO_ENABLED='1'
GOMOD='/dev/null'
GOWORK=''
CGO_CFLAGS='-O2 -g'
CGO_CPPFLAGS=''
CGO_CXXFLAGS='-O2 -g'
CGO_FFLAGS='-O2 -g'
CGO_LDFLAGS='-O2 -g'
PKG_CONFIG='pkg-config'
GOGCCFLAGS='-fPIC -m64 -pthread -Wl,--no-gc-sections -fmessage-length=0 -ffile-prefix-map=/tmp/go-build3470374486=/tmp/go-build -gno-record-gcc-switches'
```

### What did you do?

Attempted to use [`ReadOptionalASN1Boolean`](https://pkg.go.dev/golang.org/x/crypto/cryptobyte#String.ReadOptionalASN1Boolean) to read the `onlyContainsCACerts` and `onlyContainsUserCerts` fields of a CRL's `IssuingDistributionPoint` extension.

You can see my attempt, failure, and resulting investigation [here](https://github.com/letsencrypt/boulder/pull/7225). I also have a [Go playground](https://go.dev/play/p/TEVXkYVLVsg) which minimally reproduces the situation:

```go
func example(extBytes []byte) error {
	idpExt := cryptobyte.String(extBytes)
	if !idpExt.ReadASN1(&idpExt, asn1.SEQUENCE) {
		return errors.New("Failed to read IssuingDistributionPoint distributionPoint")
	}

	var onlyContainsCACerts bool
	onlyContainsCACertsTag := asn1.Tag(2).ContextSpecific()
	if !idpExt.ReadOptionalASN1Boolean(&onlyContainsCACerts, onlyContainsCACertsTag, false) {
		return errors.New("Failed to read IssuingDistributionPoint onlyContainsCACerts")
	}

	return nil
}

func main() {
	// IMPLICIT encoding
	// Sequence of length 3, containing...
	// Context-specific tag 2 of length 1, containing...
	// A byte of all ones (the representation of the boolean TRUE).
	idpExtImplicit := []byte{0x30, 0x03, 0x82, 0x01, 0xFF}
	log.Println(example(idpExtImplicit))

	// EXPLICIT encoding
	// Sequence of length 5, containing...
	// Context-specific tag 2 of length 3, containing...
	// Boolean of length 1, containing...
	// A byte of all ones (the representation of the boolean TRUE).
	idpExtExplicit := []byte{0x30, 0x05, 0x82, 0x03, 0x01, 0x01, 0xFF}
	log.Println(example(idpExtExplicit))
}
```

### What did you expect to see?

I expected ReadOptionalASN1Boolean to succeed when reading the onlyContainsCACerts field. That field is an optional boolean, so ReadOptionalASN1Boolean should successfully read it.

### What did you see instead?

ReadOptionalASN1Boolean fails when attempting to read the onlyContainsCACerts field. This is because ReadOptionalASN1Boolean _only works for explicitly-tagged fields_. It's expecting to get a five-byte sequence, as in my second example above: a context-specific tag and length, wrapping a normal boolean tag, length, and value.

But the IssuingDistributionPoint extension is defined (in both X.509 and RFC 5280) in [an ASN.1 module that sets DEFINITIONS IMPLICIT TAGS](https://datatracker.ietf.org/doc/html/rfc5280#appendix-A.2), so these fields are implicitly-tagged, not explicitly-tagged. They're only correctly encoded as a three bytes: a context-specific tag and length followed immediately by the single-byte boolean itself.

There are lots of IMPLICIT fields in x509, so it's somewhat concerning that the cryptobyte parser only has facilities for parsing EXPLICIT fields.

**Comment by thanm** [comment]
Thanks for the report.

**Comment by thanm** [comment]
@FiloSottile @rolandshoemaker @golang/security per owners

**Comment by rolandshoemaker** [comment]
I don't think we want to add transparent support for IMPLICIT values, if you expect to parse an EXPLICIT field, but get an IMPLICIT field, that seems like it should be an error (as it is now).

We've mostly avoided adding methods for all of the IMPLICIT variations of the parsing methods, since we've not really run into them all that much as of yet. The one exception to this so far is AddASN1Int64WithTag/ReadASN1Int64WithTag.

We could add ReadASN1BooleanWithTag/ReadOptionalASN1BooleanWithTag (...or something with a better name), but I feel like going down this path risks the combinatorial explosion of adding IMPLICIT versions of _everything_ we currently have.

I cannot form the exact API in my mind right now, but it seems like there is perhaps a nicer composable style API we could implement here that would allow us to just provide all of the pieces and allow the user to build the thing they want (without having to re-implement optional semantics etc).

**Comment by aarongable** [comment]
> I don't think we want to add transparent support for IMPLICIT values, if you expect to parse an EXPLICIT field, but get an IMPLICIT field, that seems like it should be an error (as it is now).

Absolutely agreed, parsing explicit vs implicit should be segregated somehow so you don't accidentally get the wrong thing.

> it seems like there is perhaps a nicer composable style API we could implement

Thinking out loud: What if the asn1 subpackage's tag type had a facility for this?

It could become a struct, rather than just a bare uint8. From there the simple solution would be to have it carry an explicit vs implicit boolean, and methods in cryptobyte could change their behavior depending on that. But perhaps an even more exciting solution would be something like
```
asn1.Tag(2).ContextSpecific() // produces 0x82
asn1.Tag(2).ContextSpecific().Explicit(asn1.BOOLEAN) // produces 0x82 <placeholder> 0x01
```
and then the cryptobyte methods would make sure that the _whole_ tag (both layers of it, in the explicit case) matches, and would use .Peek() to ensure that the next length byte is exactly 2 less than the length byte skipped over by the `<placeholder>`.

**Comment by rolandshoemaker** [comment]
Yeah Tag would probably make the most sense, although we've somewhat backed ourselves into a corner by making it just a typed uint8, so there isn't really an obvious place to hide the relevant information.

**Comment by aarongable** [comment]
Maybe the cryptobyte methods are updated to take a Tag interface, the uint8 Tag type is given a dummy method to implement that interface, and the `.Explicit(inner asn1.Tag)` method returns a different type which also implements that interface?

**Comment by rolandshoemaker** [comment]
Coming back to this, the API design of this package is clearly not ideal, and isn't particularly conducive to making these kinds of changes cleanly. There are a number of significant changes I'd like to make, but I think those need to be put off until we have the resources to do a v2 of the package.

Until we can completely overhaul the package, I think tolerating the combinatorial explosion in the number of methods is _mostly acceptable_. It's not exactly pretty, but this is an explicit parser, and having a ton of methods that just do _what they say_ is probably fine.

In terms of a concrete proposal, essentially for each of the existing `{Add,Read}ASN1{type}`, we'll add a `{Add,Read}ASN1{type}WithTag` and `{Add,Read}OptionalASN1{type}WithTag` method that support parsing the implicitly tagged types. We could also do this in a limited manner (i.e. start by only adding support for bools). Ideally we should be able to significantly reuse code in the implementation, but this is likely to still be somewhat painful.

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://go.dev/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by rolandshoemaker** [comment]
For just the Boolean type, the new APIs would be:

```
func (b *Builder) AddASN1BooleanWithTag(v bool, tag asn1.Tag)

func (s *String) ReadASN1BooleanWithTag(out *bool, tag asn1.Tag) bool
func (s *String) ReadOptionalASN1BooleanWithTag(out *bool, tag asn1.Tag, defaultValue bool) bool
```

Note `ReadOptionalASN1BooleanWithTag` has the same signature as the existing `ReadOptionalASN1Boolean`, but would work for implicitly tagged fields, rather than explicitly tagged fields.

There is an open question of whether we want to implement these APIs for the rest of the ASN1 types (BitString, Bytes, Element, Enum, GeneralizedTime, Integer, ObjectIdentifier, and UTCTime). Unless there is a current need for them (@aarongable have you run into this same problem for the other types?), I think it's reasonable to leave them out, and just implement the Boolean API for now, and then revisit this again for other types when they are needed.

**Comment by aarongable** [comment]
This API makes sense to me.

I don't think there's a strong argument for implementing the same for all types -- for example, cryptobyte currently only has "ReadOptionalASN1Foo" methods for Boolean, Integer, and OctetString, so there's prior art for not providing every possible permutation of a method.

That said, RFC 5280 does have examples of some of those types:
- The GeneralSubtrees types (used in NameConstraints extensions) has "minimum" and "maximum" fields, each of which are implicitly-tagged optional INTEGERs.
- The DistributionPoint type (used in CRLDistributionPoints extensions) has a "reasons" field which is an implicitly-tagged optional BIT STRING.
- The top-level "issuerUniqueID" and "subjectUniqueID" fields on a TBSCertificate are also both implicitly-tagged optional BIT STRINGs (but they have been profiled out of the modern PKI).
- The AuthorityKeyIdentifier type has the "keyIdentifier" field which is an implicitly-tagged optional OCTET STRING. (The go x509 lib reads this field today using a combo of PeekASN1Tag + ReadASN1, because ReadOptionalASN1OctetString doesn't work here.)

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://go.dev/s/proposal-status#likely-accept)**.
— rsc for the proposal review group


The new API is in https://github.com/golang/go/issues/64811#issuecomment-1944446920.

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://go.dev/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group


The new API is in https://github.com/golang/go/issues/64811#issuecomment-1944446920.

**Comment by pgporada** [comment]
@rolandshoemaker I have an implementation [here](https://github.com/letsencrypt/boulder/pull/7064/files#diff-dc28ebf133c75f95197326191899c7a85dc006a960c86481a7c4f24e0469922eR80-R131) if it helps.
